base/src/com/google/idea/blaze/base/sync/LockCriticalSection.kt

```diff
@@ -0,0 +1,55 @@
+package com.google.idea.blaze.base.sync
+
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.runBlockingCancellable
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.milliseconds
+
+/**
+ * Encapsulates the data into [hidden] and allows retriable access to it
+ * with cancellation checks
+ */
+class LockCriticalSection<Hidden : Any>(private val hidden: Hidden) {
```

Comment:
Since we only use this class for the SyncCache and it's only ever instantiated for a HashMap, we could drop some of the generic parameters in favor of specialization?